src/api/java/mcp/mobius/waila/api/component/NamedItemListComponent.java

```diff
@@ -0,0 +1,66 @@
+package mcp.mobius.waila.api.component;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import mcp.mobius.waila.api.ITooltipComponent;
+import mcp.mobius.waila.api.__internal__.ApiSide;
+import net.minecraft.client.gui.GuiGraphics;
+import net.minecraft.network.chat.Component;
+import net.minecraft.world.item.ItemStack;
+
+/**
+ * Component that renders items with their names.
+ */
+@ApiSide.ClientOnly
+public class NamedItemListComponent implements ITooltipComponent {
+
+    private static final NamedItemComponent MORE = new NamedItemComponent(ItemStack.EMPTY.setHoverName(Component.literal("...")));;
+
+    public NamedItemListComponent(List<ItemStack> items, int maxHeight) {
+        this.items = items;
+        this.maxHeight = maxHeight;
+    }
+
+    private final List<ItemStack> items;
+    private final int maxHeight;
+
+    private List<NamedItemComponent> getComponents() {
+        List<NamedItemComponent> components = new ArrayList<>();
```

Comment:
```suggestion
        List<NamedItemComponent> components = new ArrayList<>(items.size() + 1);
```
Since we know the size, we can specify the initial capacity to avoid growing the internal array.